### DISCUSSION TITLE
Disable "boot.py" after execution once

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#9602] [General]
Hello everybody,

I want that my application runs "boot.py" once (set up my ESP32 in default mode wifi, variables in non-volatile memory, etc) and then no more. And if a PC user order via Wifi to enable again the do it again to reset .

In other words: Start my app in default parameters, and then, don't execute "boot.py" anymore and only when "main.py" orders, execute "boot.py" again to set "factory" restore defaults.

I've been searching for boot modes but I think it's only possible power up the device in safe mode that ignores both files "boot.py" and "main.py·.

Could anybody help me?

Thank you very much!!

**Comment by dhalbert** [comment]
1. You could have boot.py write something in flash that indicates it ran already - maybe a version number?
2. `boot.py` could rename itself as part of what it does.

**Reply by nastiliano** [reply]
The second option it seems very interesting! 
If I could execute "boot.py" and at the end rename it it wouldn't be executed any more.
And then in "main.py" rename it again to "boot.py". It would be like a factory default.

How could I rename the file on the go?? (Sorry I'm a novice in Python/Micropython)

Thank you!!

**Reply by robert-hh** [reply]
import os
os.rename(oldname, newname)

**Reply by nastiliano** [reply]
This weekend I'm gonna try it! Thank you very much!

**Reply by nastiliano** [reply]
SOLVED. It works perfectly, thanks a lot!

**Comment by sosi-deadeye** [comment]
The **boot.py** could rename itself.
```
import os

# code of boot.py

# at the end
os.rename("boot.py", "_boot.py")
```

Another method could be to touch a file (_from somewhere_) and checking in **boot.py** if this files **exists**. If the file exists, skip the code in boot.py.
```
# function definitions
def main():
    """Calling all required functions in boot.py"""

skip_boot_file = "skip_boot"
if not any(file for file in os.listdir("/") if file == skip_boot_file):
    main()
```

**Reply by nastiliano** [reply]
This weekend I'm gonna try it! Thank you very much!

**Reply by robert-hh** [reply]
`OSError: [Errno 2] ENOENT` means: file not found. So the question is: what is at line 24, and which file is missing?

**Reply by robert-hh** [reply]
If boot.py does not exist, it cannot be renamed, and you get the ENOENT error. For that case, the rename may be placed into a try/except clause:

    import os
    try:
        os.rename("boot.py", "_boot.py")
    except:
        pass

**Reply by nastiliano** [reply]
SOLVED. It works perfectly, thanks a lot!

**Reply by robert-hh** [reply]
It may not be a good choice to rename boot.py to _boot.py, because before boot.py the startup code runs _boot.py, which is frozen into the flash. It looks for _boot.py as frozen script, so a file call _boot.py will not be executed. So it could only cause confusion for humans.

**Reply by nastiliano** [reply]
Ok, got it!! I changed "_boot.py" to "no_boot.py". Thanks a lot
